### PR TITLE
pythonPackages.slither-analyzer: fixed dependencies (+1 package)

### PR DESCRIPTION
**Pull request by jerengie**
###### Motivation for this change
`pythonPackage.slither-analyzer` can not be built at this moment.
This is due to missing / unlisted dependencies in the package formula.
This PR provides a fix for this and updates the formula to include the existing `pythonPackage.pysha3` package.
Additionally, another dependency, `crytic-compile`, is packaged as `pythonPackage.crytic-compile` in this PR.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/crytic-compile/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    sha256 = "32c30a9c4e6d2122fe01b21648127bca949c43402ed5a91c10bf3aa34a53731a";
+  };
+  
+  nativeBuildInputs = [ makeWrapper ];
+  propagatedBuildInputs = [ pysha3 setuptools ];
+
+  postFixup = ''
+    wrapProgram $out/bin/crytic-compile \
+      --prefix PATH : "${lib.makeBinPath [ solc ]}"
+  '';
```
Comment (jonringer) [inline pkgs/development/python-modules/crytic-compile/default.nix]: does the package also shell out to solc? if it does, then this wont work if someone imports this as a package.

Comment (jerengie) [reply pkgs/development/python-modules/crytic-compile/default.nix]: If used as a standalone tool, yes. 
I am not aware of any nix package that depends on this besides `slither-analyzer`, which also includes this `postFixup`.

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -3349,6 +3349,12 @@
     githubId = 1608697;
     name = "Jens Binkert";
   };
+  jerengie = {
+    email = "me@jerengie.de";
+    github = "jerengie";
+    githubID = 1221065;
+    name = "Jens-Rene Giesen";
+  };
```
Comment (jonringer) [inline maintainers/maintainer-list.nix]: this should be a separate commit as well:
```
maintainers: add jerengie
```

**Review by jonringer — COMMENTED** [review]
there should be 3 commits:
```
maintainers: add jerengie
pythonPackages.slither-analyzer: fix dependencies
pythonPackages.crytic-compile: init at 0.1.6
```

File: pkgs/development/python-modules/crytic-compile/default.nix (hunk 1 above)
Comment (asymmetric) [reply pkgs/development/python-modules/crytic-compile/default.nix]: > does the package also shell out to solc? if it does, then this wont work if someone imports this as a package.

@jonringer why is that?

File: pkgs/development/python-modules/crytic-compile/default.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  meta = with lib; {
+    description = "Abstraction layer for smart contract build systems";
+    longDescription = ''
+      Crytic-compile is a library to help smart contract compilation.
+      It is used in many Crytic tools, e.g. Slither.
+    '';
+    homepage = https://github.com/crytic/crytic-compile;
```
Comment (asymmetric) [inline pkgs/development/python-modules/crytic-compile/default.nix]: The URL should be double quoted, the URL type is going away.

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -1552,6 +1552,8 @@ in
 
   crudini = callPackage ../tools/misc/crudini { };
 
+  crytic-compile = with python3Packages; toPythonApplication crytic-compile;
```
Comment (asymmetric) [inline pkgs/top-level/all-packages.nix]: Does it make sense to have this in `all-packages`? Seems like it will mostly be used as a dependency of other packages, so not sure users will want to install it directly very often.

**Review by asymmetric — COMMENTED** [review]
@jerengie thanks, this looks good! (modulo the changes outlined by @jonringer)

Would you consider adding yourself to the maintainers? I'm not using this anymore.

File: maintainers/maintainer-list.nix (hunk 2 above)
Comment (jerengie) [reply maintainers/maintainer-list.nix]: Thank you, I corrected this.

File: pkgs/development/python-modules/crytic-compile/default.nix (hunk 3 above)
Comment (jerengie) [reply pkgs/development/python-modules/crytic-compile/default.nix]: > @jerengie thanks, this looks good! (modulo the changes outlined by @jonringer)
> Would you consider adding yourself to the maintainers? I'm not using this anymore.

Thank you. Sure, I now added myself to the list of maintainers for `slither-analyzer`.

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (jerengie) [reply pkgs/top-level/all-packages.nix]: Thanks for this suggestion.
To be honest, I have no idea if users would like to install `crytic-compile` in a standalone way.
But if desired, it is nice to have the opportunity to do so, isn't it?

File: pkgs/development/python-modules/crytic-compile/default.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/development/python-modules/crytic-compile/default.nix]: because you only wrap the executable, and the python package package has no knowledge of it.

If i had another package which has `propagatedBuildInputs = [ cryptic-compile ];` as a dependenccy, then when I call `$ my_package_exe` and it eventually calls cryptic-compile, which will do something to the effect of `subprocess.check_call("solc" ...)`, then solc won't be present.

If you want the package to do the right thing, you need to patch the call sights inside of only wrapping the entry_points.

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (FRidh) [reply pkgs/top-level/all-packages.nix]: No we shouldn't pollute the top-level unnecessarily.

**Review by FRidh — CHANGES_REQUESTED** [review]
- patch code to executable
- remove top-level attribute

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (asymmetric) [reply pkgs/top-level/all-packages.nix]: @jerengie it will still be available under `python3Packages`, so it can still be installed, just not from the top-level namespace.

@FRidh the manual does say though:

> A distinction is made between applications and libraries, however, sometimes a package is used as both. In this case the package is added as a library to `python-packages.nix` and as an application to `all-packages.nix`.

Comment (FRidh) [reply pkgs/top-level/all-packages.nix]: @asymmetric yes but if the author of the PR isn't going to use it like that to begin with, should we then add it just for potential users?

Comment (asymmetric) [reply pkgs/top-level/all-packages.nix]: I guess not. Maybe the manual should mention that the package *may* be added to `all-packages.nix`.

**Comment by jonringer** [comment]
stale, and conflicts

I'm closing this PR due to lack of activity, if the PR author would like to pursue this matter again, please re-open the PR and address the concerns.
